File: src/main/java/org/modelingvalue/nelumbo/syntax/TokenType.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.syntax;

import static org.modelingvalue.nelumbo.syntax.TokenType.Flag.CONTINUES_ON_NEXT_LINE;
import static org.modelingvalue.nelumbo.syntax.TokenType.Flag.LAYOUT;
import static org.modelingvalue.nelumbo.syntax.TokenType.Flag.NOT_MATCHED;
import static org.modelingvalue.nelumbo.syntax.TokenType.Flag.SKIP;
import static org.modelingvalue.nelumbo.syntax.TokenType.Flag.VARIABLE_CONTENT;

import java.util.Arrays;
import java.util.EnumSet;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public enum TokenType {
    SINGLEQUOTE("'"), //
    SEMICOLON(";"), //
    COMMA(",", CONTINUES_ON_NEXT_LINE), //
    LEFT("[\\(\\[\\{]", CONTINUES_ON_NEXT_LINE, VARIABLE_CONTENT), //
    RIGHT("[\\)\\]\\}]", VARIABLE_CONTENT), //
    STRING("\"([^\"\\\\]|\\\\[\\s\\S])*\"", VARIABLE_CONTENT), //
    DECIMAL("-?[0-9]+\\.[0-9]+", VARIABLE_CONTENT), //
    NUMBER("-?[0-9]+(#[0-9a-zA-Z]+)?", VARIABLE_CONTENT), //
    NAME("[a-zA-Z_][0-9a-zA-Z_]*", VARIABLE_CONTENT), //
    META_OPERATOR("<(\\(|\\)|\\)\\?|\\)\\*|\\)\\+|\\,|\\|)>", VARIABLE_CONTENT), //
    END_LINE_COMMENT("//[^\\v]*", SKIP, VARIABLE_CONTENT), //
    IN_LINE_COMMENT("/\\*.*?(?:\\*/|\\z)", SKIP, VARIABLE_CONTENT), //
    OPERATOR("(?!//)[~!@#$%^&*=+|:<>.?/-]+", CONTINUES_ON_NEXT_LINE, VARIABLE_CONTENT), //
    NEWLINE("\\R", CONTINUES_ON_NEXT_LINE, LAYOUT), //
    HSPACE("\\h+", SKIP, LAYOUT), //
    ERROR(".", VARIABLE_CONTENT), //
    //================ rest is not actually matched:
    BEGINOFFILE, //
    ENDOFFILE, //
    ENDOFLINE, //
    VARIABLE, //
    KEYWORD, //
    TYPE, //
    ;

    // Combined patterns for efficient matching - built once at class load
    private static final Pattern     COMBINED_FULL_MATCH_PATTERN;
    private static final Pattern     COMBINED_PATTERN;
    private static final TokenType[] COMBINED_PATTERN_LOOKUP;

    static {
        StringBuilder sb     = new StringBuilder();
        String        sep    = "";
        int           group  = 1;
        TokenType[]   lookup = new TokenType[TokenType.values().length + 1]; // index 0 unused (groups start at 1)
        for (TokenType tt : TokenType.values()) {
            if (!tt.isNotMatched()) {
                // Convert internal capturing groups to non-capturing (but not escaped literal parens like \()
                String p = tt.pattern.pattern().replaceAll("(?<!\\\\)\\((?!\\?)", "(?:");
                sb.append(sep).append("(").append(p).append(")");
                sep           = "|";
                lookup[group] = tt;
                group++;
            }
        }
        String fullRegexp = sb.toString();
        //noinspection RegExpUnnecessaryNonCapturingGroup
        COMBINED_FULL_MATCH_PATTERN = Pattern.compile("^(?:" + fullRegexp + ")$", Pattern.DOTALL);
        COMBINED_PATTERN            = Pattern.compile(fullRegexp, Pattern.MULTILINE | Pattern.DOTALL);
        COMBINED_PATTERN_LOOKUP     = Arrays.copyOf(lookup, group);
    }

    public enum Flag {
        SKIP,                   // indicates a non semantic part that may be ignored by the parser
        CONTINUES_ON_NEXT_LINE, // indicates that a NEWLINE token after this token is to be ignored when parsing
        NOT_MATCHED,            // indicates that this token type is not actually matched by the lexer
        VARIABLE_CONTENT,       // indicates that this token type has a variable content
        LAYOUT                  // indicates that this token type is layout and should be ignored by the parser
    }

    private final Pattern pattern;
    private final boolean skip;
    private final boolean continuesOnNextLine;
    private final boolean layout;
    private final boolean variableContent;
    private final boolean notMatched;

    TokenType() {
        this("", LAYOUT, NOT_MATCHED);
    }

    TokenType(String regexp, Flag... flags) {
        this.pattern = Pattern.compile(regexp, Pattern.MULTILINE | Pattern.DOTALL);
        EnumSet<Flag> flagset = flags.length == 0 ? EnumSet.noneOf(Flag.class) : EnumSet.of(flags[0], flags);
        this.skip                = flagset.contains(SKIP);
        this.continuesOnNextLine = flagset.contains(CONTINUES_ON_NEXT_LINE);
        this.layout              = flagset.contains(LAYOUT);
        this.variableContent     = flagset.contains(VARIABLE_CONTENT);
        this.notMatched          = flagset.contains(NOT_MATCHED);
    }

    public Pattern pattern() {
        return pattern;
    }

    public boolean isSkip() {
        return skip;
    }

    public boolean isContinuesOnNextLine() {
        return continuesOnNextLine;
    }

    public boolean isLayout() {
        return layout;
    }

    public boolean isVariableContent() {
        return variableContent;
    }

    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
    public boolean isNotMatched() {
        return notMatched;
    }

    public boolean matches(String text) {
        return pattern.matcher(text).matches();
    }

    public static TokenType of(String text) {
        Matcher m = COMBINED_FULL_MATCH_PATTERN.matcher(text);
        if (m.matches()) {
            int group = findGroup(m);
            if (group != 0) {
                return COMBINED_PATTERN_LOOKUP[group];
            }
        }
        return null;
    }

    private static int findGroup(Matcher m) {
        for (int group = 1; group < COMBINED_PATTERN_LOOKUP.length; group++) {
            if (m.start(group) >= 0) {
                return group;
            }
        }
        return 0;
    }

    public static TokenMatcher getMatcher(String input) {
        return new TokenMatcher(input);
    }

    public static class TokenMatcher {
        private final Matcher   matcher;
        private final int       inputLength;
        private       String    matchedText;
        private       TokenType matchedType;

        private TokenMatcher(String input) {
            this.matcher     = COMBINED_PATTERN.matcher(input);
            this.inputLength = input.length();
        }

        public boolean hasMore() {
            if (inputLength <= matcher.regionStart()) {
                return false;
            }
            if (matcher.lookingAt()) {
                int group = findGroup(matcher);
                if (group != 0) {
                    matchedText = matcher.group(group);
                    matchedType = COMBINED_PATTERN_LOOKUP[group];
                    matcher.region(matcher.end(), inputLength);
                    return true;
                }
            }
            throw new IllegalStateException("TokenMatcher does not match input at: " + matcher.regionStart());
        }

        public String text() {
            return matchedText;
        }

        public TokenType type() {
            return matchedType;
        }
    }
}

File: web-ts/src/syntax/TokenType.ts
/**
 * Token types for Nelumbo lexer.
 * Ported from Java: org.modelingvalue.nelumbo.syntax.TokenType
 */

export enum TokenTypeFlag {
  SKIP = 1,                    // Non-semantic token (comments, whitespace)
  CONTINUES_ON_NEXT_LINE = 2,  // Newlines after these tokens are ignored
  VARIABLE_CONTENT = 4,        // Token content is variable/meaningful
  LAYOUT = 8,                  // Layout/formatting token
  NOT_MATCHED = 16,            // Not matched by lexer (synthetic)
}

export interface TokenTypeDef {
  name: string;
  pattern: RegExp | null;
  flags: number;
}

// Token type definitions
const tokenTypeDefs: TokenTypeDef[] = [
  // Matched tokens (in priority order - comments must come before OPERATOR)
  { name: 'SINGLEQUOTE', pattern: /^'/, flags: 0 },
  { name: 'SEMICOLON', pattern: /^;/, flags: 0 },
  { name: 'COMMA', pattern: /^,/, flags: TokenTypeFlag.CONTINUES_ON_NEXT_LINE },
  { name: 'LEFT', pattern: /^[(\[{]/, flags: TokenTypeFlag.CONTINUES_ON_NEXT_LINE | TokenTypeFlag.VARIABLE_CONTENT },
  { name: 'RIGHT', pattern: /^[)\]}]/, flags: TokenTypeFlag.VARIABLE_CONTENT },
  { name: 'STRING', pattern: /^"([^"\\]|\\[\s\S])*"/, flags: TokenTypeFlag.VARIABLE_CONTENT },
  { name: 'DECIMAL', pattern: /^-?[0-9]+\.[0-9]+/, flags: TokenTypeFlag.VARIABLE_CONTENT },
  { name: 'NUMBER', pattern: /^-?[0-9]+(#[0-9a-zA-Z]+)?/, flags: TokenTypeFlag.VARIABLE_CONTENT },
  { name: 'NAME', pattern: /^[a-zA-Z_][0-9a-zA-Z_]*/, flags: TokenTypeFlag.VARIABLE_CONTENT },
  { name: 'NEWLINE', pattern: /^(\r\n|\r|\n)/, flags: TokenTypeFlag.CONTINUES_ON_NEXT_LINE | TokenTypeFlag.LAYOUT },
  { name: 'HSPACE', pattern: /^[ \t]+/, flags: TokenTypeFlag.SKIP },
  { name: 'END_LINE_COMMENT', pattern: /^\/\/[^\r\n]*/, flags: TokenTypeFlag.SKIP | TokenTypeFlag.VARIABLE_CONTENT },
  { name: 'IN_LINE_COMMENT', pattern: /^\/\*[\s\S]*?(\*\/|$)/, flags: TokenTypeFlag.SKIP | TokenTypeFlag.VARIABLE_CONTENT },
  { name: 'OPERATOR', pattern: /^(?!\/\/|\/\*)[~!@#$%^&*=+|:<>.?/-]+/, flags: TokenTypeFlag.CONTINUES_ON_NEXT_LINE | TokenTypeFlag.VARIABLE_CONTENT },
  { name: 'ERROR', pattern: /^./, flags: 0 },

  // Synthetic tokens (not matched by lexer)
  { name: 'BEGINOFFILE', pattern: null, flags: TokenTypeFlag.NOT_MATCHED | TokenTypeFlag.LAYOUT },
  { name: 'ENDOFFILE', pattern: null, flags: TokenTypeFlag.NOT_MATCHED | TokenTypeFlag.LAYOUT },
  { name: 'ENDOFLINE', pattern: null, flags: TokenTypeFlag.NOT_MATCHED | TokenTypeFlag.LAYOUT },

  // Semantic tokens (for syntax highlighting, not matched by lexer)
  { name: 'VARIABLE', pattern: null, flags: TokenTypeFlag.NOT_MATCHED | TokenTypeFlag.VARIABLE_CONTENT },
  { name: 'KEYWORD', pattern: null, flags: TokenTypeFlag.NOT_MATCHED },
  { name: 'TYPE', pattern: null, flags: TokenTypeFlag.NOT_MATCHED },
  { name: 'META_OPERATOR', pattern: null, flags: TokenTypeFlag.NOT_MATCHED },
];

export class TokenType {
  private static readonly types: Map<string, TokenType> = new Map();
  private static readonly matchedTypes: TokenType[] = [];

  readonly name: string;
  readonly pattern: RegExp | null;
  readonly flags: number;

  private constructor(def: TokenTypeDef) {
    this.name = def.name;
    this.pattern = def.pattern;
    this.flags = def.flags;
  }

  static initialize(): void {
    for (const def of tokenTypeDefs) {
      const type = new TokenType(def);
      TokenType.types.set(def.name, type);
      if (def.pattern !== null) {
        TokenType.matchedTypes.push(type);
      }
    }
  }

  static get(name: string): TokenType {
    const type = TokenType.types.get(name);
    if (!type) {
      throw new Error(`Unknown token type: ${name}`);
    }
    return type;
  }

  static getMatchedTypes(): TokenType[] {
    return TokenType.matchedTypes;
  }

  // Common token types as static properties
  static SINGLEQUOTE: TokenType;
  static SEMICOLON: TokenType;
  static COMMA: TokenType;
  static LEFT: TokenType;
  static RIGHT: TokenType;
  static STRING: TokenType;
  static DECIMAL: TokenType;
  static NUMBER: TokenType;
  static NAME: TokenType;
  static OPERATOR: TokenType;
  static NEWLINE: TokenType;
  static HSPACE: TokenType;
  static END_LINE_COMMENT: TokenType;
  static IN_LINE_COMMENT: TokenType;
  static ERROR: TokenType;
  static BEGINOFFILE: TokenType;
  static ENDOFFILE: TokenType;
  static ENDOFLINE: TokenType;
  static VARIABLE: TokenType;
  static KEYWORD: TokenType;
  static TYPE: TokenType;
  static META_OPERATOR: TokenType;

  isSkip(): boolean {
    return (this.flags & TokenTypeFlag.SKIP) !== 0;
  }

  isContinuesOnNextLine(): boolean {
    return (this.flags & TokenTypeFlag.CONTINUES_ON_NEXT_LINE) !== 0;
  }

  isLayout(): boolean {
    return (this.flags & TokenTypeFlag.LAYOUT) !== 0;
  }

  isVariableContent(): boolean {
    return (this.flags & TokenTypeFlag.VARIABLE_CONTENT) !== 0;
  }

  isNotMatched(): boolean {
    return (this.flags & TokenTypeFlag.NOT_MATCHED) !== 0;
  }

  /**
   * Determine the token type for a given text string.
   */
  static of(text: string): TokenType {
    for (const type of TokenType.matchedTypes) {
      if (type.pattern && type.pattern.test(text)) {
        // Verify it matches the whole string
        const match = text.match(type.pattern);
        if (match && match[0] === text) {
          return type;
        }
      }
    }
    return TokenType.ERROR;
  }
}

// Initialize token types
TokenType.initialize();

// Set static properties
TokenType.SINGLEQUOTE = TokenType.get('SINGLEQUOTE');
TokenType.SEMICOLON = TokenType.get('SEMICOLON');
TokenType.COMMA = TokenType.get('COMMA');
TokenType.LEFT = TokenType.get('LEFT');
TokenType.RIGHT = TokenType.get('RIGHT');
TokenType.STRING = TokenType.get('STRING');
TokenType.DECIMAL = TokenType.get('DECIMAL');
TokenType.NUMBER = TokenType.get('NUMBER');
TokenType.NAME = TokenType.get('NAME');
TokenType.OPERATOR = TokenType.get('OPERATOR');
TokenType.NEWLINE = TokenType.get('NEWLINE');
TokenType.HSPACE = TokenType.get('HSPACE');
TokenType.END_LINE_COMMENT = TokenType.get('END_LINE_COMMENT');
TokenType.IN_LINE_COMMENT = TokenType.get('IN_LINE_COMMENT');
TokenType.ERROR = TokenType.get('ERROR');
TokenType.BEGINOFFILE = TokenType.get('BEGINOFFILE');
TokenType.ENDOFFILE = TokenType.get('ENDOFFILE');
TokenType.ENDOFLINE = TokenType.get('ENDOFLINE');
TokenType.VARIABLE = TokenType.get('VARIABLE');
TokenType.KEYWORD = TokenType.get('KEYWORD');
TokenType.TYPE = TokenType.get('TYPE');
TokenType.META_OPERATOR = TokenType.get('META_OPERATOR');
